### ISSUE TITLE
Header Component Architecture & Deployment Workflow Improvements

### ISSUE THREAD

**Issue by rhernand3z** [NYPL/dgx-header-component#3]
I've generated documentation as well as a technical diagram that outlines:

- Current Technical Diagram of the Header workflow
- How the header component is utilized across nypl.org web properties
- Our current solution to embed the Header in React Apps, Drupal, Angular Apps and 3rd party vendor websites
- Topics of discussion around improving the architecture
- Topics of discussion around improving the build/deployment workflow

[State of the Header](https://docs.google.com/a/nypl.org/presentation/d/1tmudf7ygXDeSyYiJeXijjtZJ1d0FDD9HlcOGQAldv9g/edit?usp=sharing)

CC: @natalietsch @kfriedman @EdwinGuzman @nonword @ricardoom @courtneymcgee @gkallenberg @nodanaonlyzuul @ktp242 @seanredmond @jvarghese01  @holingpoon

**Comment by kfriedman** [comment]
This really fantastic. Thanks so much for doing this.

I've been thinking about the header architecture a bit. Primarily, how can we build an architecture so the header can support _current_ needs:

- Easier updates/deploys
- Login session awareness
- Improved accessibility (especially no JS)

And, possible _future_ needs:

- Better account access (e.g. show a patron's barcode, show patron's overdue books)
- Notifications
- Fundraising
- ... and other stuff to be determined

I've come up with three approaches so far. I'm sure there are many more. There are definitely pros and cons to each approach. Here's a quick summary of what I've been able to come up with so far:

# 1. NPM Dependency (our current approach)
Include and render the header as an Node module/React component.

## Pros
- Probably the most common/standardized approach for a React app.

## Cons
- Potentially puts us in dependency hell.
- Potentially makes small updates difficult (i.e. requires super deployments for simple changes).
- Potentially makes large functionality updates difficult and risky.
- Only works server-side for React apps and therefore also relies on a server-side include solution as well.

# 2. Rendered Server-Side Include
Include the header server-side with pre-rendered HTML/CSS/JS.

## Pros
- Universal (i.e. can be included in any non-React app with a HTTP request).
- Much easier to make updates; no more super deployments.
- No dependency issues.

## Cons
- Feels "hacky" and may cause React issues.
- Potential trickiness to make it login session aware.
- Potential performance/scaleability issues.

_... and a much more radical approach that may not even be technically possible but I'll throw it out there..._

# 3. Global Reverse Proxy
Build a new reverse-proxy that will "assemble" NYPL page requests (i.e. put the header at the top and main page content below it—maybe with other components).

## Pros
- Could provide true server-side rendering on *all* NYPL properties including Encore and any future property.
- Could provide one place to include global components and functionality.
- Universal (i.e. doesn't matter what the destination property is running).
- Much easier to make updates; no more super deployments.

## Cons
- May not be technically possible.
- Uncharted territory and, therefore, unknown risks .
- Potential performance/scalability issues.

**Comment by nodanaonlyzuul** [comment]
In regards to serveside includes...

Rafael & I did some POC work on [NYPL/dgx-homepage](https://bitbucket.org/NYPL/dgx-homepage) and at first blush yes, NYPL/dgx-homepage's state manager will freak out if another React app is brought into the template. Maybe @rhernand3z has ideas about how to make it freak out less. I'm not sure if serverside includes are the best answer - but devil's advocate...we wouldn't have the collision if the header weren't React. Then again...we get a lot from it being a react app.

A really simple / maybe heavy-handed way to reduce dependency hell is to consolidate/monolith-ize all the SPAs currently depend on the header. One big repo that would serve get-a-library-card, homepage, etc. There are drawbacks for this surely - but I wonder how much each app's independence has become a liability and not an asset.

I'm having a hard time parsing out what the best approach for us is...and if these solutions are just antipatterns to having a bad deployment story.

**Comment by nodanaonlyzuul** [comment]
For posterity there's conversation about this happening here:
https://jira.nypl.org/browse/HDR-42

**Comment by EdwinGuzman** [comment]
It's been six months since we've last touched base on this issue. Since then, with help from IT and devs at Digital, we've moved all Node/React and Sinatra/Angular apps to our AWS accounts and have set up Travis for CI/CD on all those apps. That has helped tremendously in quickly deploying to the apps that include this header component as an npm package. We went from about 45+ minutes of building an artifact and deploying using Bamboo to about 10 minutes of _testing_, building, and deploying (testing before building and deploying has been a great added benefit). The time estimate is based on the time after we merge to the `master` branch and then let Bamboo/Travis take over.

From the discussion above, we've been following the `NPM Dependency` approach. The following updates and notes are based on the past six months of active development.

**Pros**
> * Probably the most common/standardized approach for a React app.

✅ 
> * Enables (forces?) us to build a CD pipeline.

I'd say _enable_ is the better wording and this has also helped us run tests before building and deploying.

**Cons**
> * Potentially puts us in dependency hell.

This is still an issue but we haven't encountered many cases where this impedes development or deployments.
> * Potentially makes small updates difficult (i.e. requires super deployments for simple changes).

This has been true. Fortunately, there has been multiple updates that have been bundled together into new versions. This makes deployments for all the apps easier.

> * Potentially makes large functionality updates difficult and risky.

Likewise, when we have many new features, or bug fixes, checking all of them off on all the apps is tedious. 
> * Only works server-side for React apps and therefore also relies on a server-side include solution as well.

This hasn't been a problem for us.

---
We haven't changed much of the development workflow and it has gone well so far. But, it is also difficult to judge because we have not had any new and big features added. 

@wlla  and I have been working on documenting where the header can be found, what version of the header can be found on those apps, and any inconsistencies. Most recent documentation can be found on:
1. List of apps and what version of the header it contains: https://docs.google.com/spreadsheets/d/1h5oAFT9H0efEgTiAM1bfqNDK83STc1qfaVCdRzUFmOI/edit?usp=sharing
2. Working documentation on Skip Navigation inconsistency, but has useful information regarding how the Header is instantiated on each app: https://docs.google.com/document/d/1M4ysO0Jg5ANZtT0fSuENqChdlQ2xvPeiniIbJDFwXgs/edit#heading=h.v1gk3fnsypx7

---
I'm fine with this workflow (# 1 described by Rafael in the comments above) as long as the documentation keeps up. I do think @ricardoom @gkallenberg @ktp242 @holingpoon should weigh in their thoughts on recent development on the header and how it went for them. I will be moving most of the documentation for the header, once it's finalized and better, into this repo and the [Header App](https://github.com/NYPL/nypl-dgx-react-header/) repo.

If anyone has any more thoughts on improving the workflow for the header dependency, development, or deployment, feel free to contribute to this. _I think_ it's safe to say that we will be moving along with the npm package approach for now.

**Comment by holingpoon** [comment]
Do we still have a copy of [State of the Header](https://docs.google.com/a/nypl.org/presentation/d/1tmudf7ygXDeSyYiJeXijjtZJ1d0FDD9HlcOGQAldv9g/edit?usp=sharing) somewhere? As part of a bigger conversation, we need to park documentation and diagrams at places that would not get erased after our Active Directory credentials are revoked. I highly recommend Markdown in a `docs` or `documentation` directory in the code repo so at least we can cross-reference in other repos. From a standpoint of the information provided by the list of Google Docs, Google Spreadsheets, and Google Slides, most if not all of the info, in my opinion, should really be part of the repo's documentation. 

In general, I think the benefits outweigh the risks with CI/CD. To be honest, my expectation of what CI/CD can do for us is too high. I was hoping that CI/CD would improve our workflow so much, that a simple push to the `development`, `qa`, and `production` branches would deploy everywhere. 

In reality, we still need to do non-trivial amount coordination for header deployment since we need to update header version on every app that uses the header, which right now stands around 18-20 apps. If more than one dev is working on the header, we need to coordinate branch merging strategies so the merge flow, version numbers, and tagging are consistent, but that problem is not unique to the header component, the header app, or even to NYPL. I believe as long as we are diligent, we should be okay.

**Comment by EdwinGuzman** [comment]
Agreed, the workflow could be better but at least it's better than months ago. We actually only have 9 apps that need Header component version bumps and one of those apps affects about 5 others.

I'd be happy with closing this ticket after I port over documentation from docs and spreadsheets to this repo as well as to the [Header app](https://github.com/NYPL/nypl-dgx-react-header) repo.
